resolver/addressable_resolver.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resolver
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+
+	"knative.dev/eventing/pkg/reconciler/names"
+	"knative.dev/pkg/apis"
+	pkgapisduck "knative.dev/pkg/apis/duck"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+	apisv1alpha1 "knative.dev/pkg/apis/v1alpha1"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/tracker"
+
+	"knative.dev/pkg/injection/clients/dynamicclient"
+)
+
+// ForDestination resolves a Destination into a URI.
+// This is a helper component for resources that references other resources' addresses.
+type ForDestination struct {
+	tracker         tracker.Interface
+	informerFactory pkgapisduck.InformerFactory
+}
+
+// NewForDestination creates and initializes a new resolver.ForDestination.
+func NewForDestination(ctx context.Context, callback func(string)) *ForDestination {
+	ret := &ForDestination{}
+
+	ret.tracker = tracker.New(callback, controller.GetTrackerLease(ctx))
+	ret.informerFactory = &pkgapisduck.CachedInformerFactory{
+		Delegate: &pkgapisduck.EnqueueInformerFactory{
+			Delegate: &pkgapisduck.TypedInformerFactory{
+				Client:       dynamicclient.Get(ctx),
+				Type:         &duckv1beta1.AddressableType{},
+				ResyncPeriod: controller.GetResyncPeriod(ctx),
+				StopChannel:  ctx.Done(),
+			},
+			EventHandler: controller.HandleAll(ret.tracker.OnChanged),
+		},
+	}
+
+	return ret
+}
+
+// GetURI resolves a Destination into a URI string.
+func (r *ForDestination) GetURI(dest apisv1alpha1.Destination, parent interface{}) (string, error) {
+	// Prefer resolved object reference + path, then try URI + path, honoring the Destination documentation
+	if dest.ObjectReference != nil {
+		url, err := r.resolveObjectReference(dest.ObjectReference, parent)
```

Comment:
We need to expose resolveObjectReference as a top level option, perhaps two objects

Reply:
You can always

```go
uri, err := r.GetURI(apis.Destination{ObjectReference: myRefHere}, myParent)
```
but that's verbose and costs some allocation.

Perhaps it would be nice to have `ResolveDestination` and `ResolveObjectReference` instead of `GetURI`.

That changes the language, but I think I like `URIResolver.ResolveDestination` - the input/output relationship there is more clear.

Reply:
some parts of the code will only want to deal in addressable refs